cub/util_type.cuh

```diff
@@ -273,9 +274,47 @@ struct NullType
 template <int A>
 struct Int2Type
 {
-   enum {VALUE = A};
+    enum {VALUE = A};
 };
 
+/**
+ * \brief Allows algorithms that take a value as input to take a future value that is not computed yet at launch time.
+ *
+ * Note that it is user's responsibility to ensure that the result will be ready before use via external synchronization
+ * or stream-ordering dependencies.
+ */
+template <typename T, typename IterT = T*>
+struct FutureValue
+{
+    using value_type = T;
+    using iterator_type = IterT;
+    IterT iter;
+    __host__ __device__ __forceinline__ FutureValue(IterT iter):iter(iter) {}
+    __host__ __device__ __forceinline__ operator T() {
+        return *iter;
+    }
+};
+
+template <typename T, typename IterT = T*>
+struct InputValue
+{
+    using value_type = T;
+    using iterator_type = IterT;
+    bool is_future;
+    union
+    {
+        FutureValue<T, IterT> future_value;
+        T immediate_value;
+    };
+    __host__ __device__ __forceinline__ operator T() {
+        if (is_future) {
+            return future_value;
+        }
+        return immediate_value;
+    }
+    __host__ __device__ __forceinline__ InputValue(T immediate_value): is_future(false), immediate_value(immediate_value) {}
```

Comment:
Same notes as above:
- These constructors should be marked `explicit`.
- Use a different name for either the argument or member to avoid shadowing warnings. I'd suggest prefixing the member variables with `m_` since we're considering adopting this style convention "soon".
- Make the internal member variable + union private, since users shouldn't directly interact with them.

Reply:
fixed